src/main/java/net/glowstone/util/nbt/CompoundTag.java

```diff
@@ -471,23 +471,19 @@ public void putIntArray(String key, int... value) {
     /**
      * Adds or replaces a list subtag, converting the list entries to tags.
      *
+     * @param <V> the list elements' Java type
      * @param key the key to write to
      * @param type the list elements' tag type
      * @param value the list contents, as objects to convert to tags
-     * @param <V> the list elements' Java type
+     * @param tagCreator a function that will convert each V to an element tag
      */
-    public <V> void putList(String key, TagType type, List<V> value) {
-        // the reflection here is really gross but I'm not sure a good way around it
-        try {
-            Constructor<? extends Tag> constructor = type.getConstructor();
-            List<Tag> result = new ArrayList<>(value.size());
-            for (V item : value) {
-                result.add(constructor.newInstance(item));
-            }
-            put(key, new ListTag<>(type, result));
-        } catch (ReflectiveOperationException e) {
-            throw new IllegalArgumentException("Unable to create list of type " + type, e);
+    public <V> void putList(String key, TagType type, List<V> value,
+            Function<? super V, ? extends Tag> tagCreator) {
```

Comment:
Why not store the method reference in the enum?

Reply:
That would create type-safety issues, because an enum can't be generic, so I'm creating `put*List` wrappers instead.